### PR TITLE
[INLONG-8038][Sort] Optimize MySQL CDC chunk splitting

### LINKED ISSUE #8038: [Feature][Sort] Optimize MySQL CDC chunk splitting logic
### Description

[Feature][Sort] Optimize MySQL CDC chunk splitting logic

### Use case

_No response_

### Are you willing to submit PR?

- [X] Yes, I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by e-mhui**
### Prepare a Pull Request

[INLONG-8038][Sort] Optimize MySQL CDC chunk splitting logic

- Fixes #8038 

### Motivation

When the data volume grows rapidly and the snapshot phase takes a long time, the data volume of the last chunk may be very large, leading to OOM. Therefore, it is recommended to read the last chunk first to prevent OOM.
Because these splits are assigned to readers in order, we only need to adjust the order of splits by moving the last split to the front.

```java
  // assign splits in order 
  Optional<MySqlSplit> split = splitAssigner.getNext();
  if (split.isPresent()) {
      final MySqlSplit mySqlSplit = split.get();
      context.assignSplit(mySqlSplit, nextAwaiting);
      awaitingReader.remove();
      LOG.info("Assign split {} to subtask {}", mySqlSplit, nextAwaiting);
  } else {
      // there is no available splits by now, skip assigning
      wakeupBinlogReaderIfNeed();
      break;
  }
```
code link: https://github.com/apache/inlong/blob/370815119d7d8898bbf13febc5b70628893ef43f/inlong-sort/sort-flink/sort-flink-v1.13/sort-connectors/mysql-cdc/src/main/java/org/apache/inlong/sort/cdc/mysql/source/enumerator/MySqlSourceEnumerator.java#LL200C1-L210C14

### Modifications

Move the last snapshot split to the front of the remaining splits.

```java
synchronized (lock) {
    remainingSplits.addAll(schemaLessSnapshotSplits);
    int size = schemaLessSnapshotSplits.size();
    // move the last snapshot split to the front of the remaining splits.
    remainingSplits.add(0, schemaLessSnapshotSplits.get(size - 1));
    remainingSplits.addAll(schemaLessSnapshotSplits.subList(0, size - 1));
    remainingTables.remove(nextTable);
    addAlreadyProcessedTablesIfNotExists(nextTable);
    lock.notify();
}
```

### CODE REVIEW

**Comment by gong** [comment]
@e-mhui Hi, Will be binlog offset effected?

**Comment by e-mhui** [comment]
> @e-mhui Hi, Will be binlog offset effected?

It will not affect binlog offset, because the binlog offset will start from the smallest high watermark in the chunk splits.
image

File: inlong-sort/sort-flink/sort-flink-v1.13/sort-connectors/mysql-cdc/src/main/java/org/apache/inlong/sort/cdc/mysql/source/assigners/MySqlSnapshotSplitAssigner.java [hunk 1]
```diff
@@ -228,7 +228,10 @@ private void startAsynchronouslySplit() {
                                                 .map(MySqlSnapshotSplit::toSchemaLessSnapshotSplit)
                                                 .collect(Collectors.toList());
                                 synchronized (lock) {
-                                    remainingSplits.addAll(schemaLessSnapshotSplits);
+                                    int size = schemaLessSnapshotSplits.size();
```
Comment (EMsnap) [inline inlong-sort/sort-flink/sort-flink-v1.13/sort-connectors/mysql-cdc/src/main/java/org/apache/inlong/sort/cdc/mysql/source/assigners/MySqlSnapshotSplitAssigner.java:231]: I guess the steps can be held inside one method. ?

Comment (e-mhui) [reply inlong-sort/sort-flink/sort-flink-v1.13/sort-connectors/mysql-cdc/src/main/java/org/apache/inlong/sort/cdc/mysql/source/assigners/MySqlSnapshotSplitAssigner.java]: done.

**Review by EMsnap — APPROVED** [review]

**Review by dockerzhang — APPROVED** [review]

File: inlong-sort/sort-flink/sort-flink-v1.13/sort-connectors/mysql-cdc/src/main/java/org/apache/inlong/sort/cdc/mysql/source/assigners/MySqlSnapshotSplitAssigner.java [hunk 2]
```diff
@@ -397,6 +397,14 @@ private void closeExecutorService() {
         }
     }
 
+    private void addNewlyAddedSplits(List<MySqlSchemalessSnapshotSplit> schemaLessSnapshotSplits) {
+        int size = schemaLessSnapshotSplits.size();
+        // move the last snapshot split to the front of the remaining splits to prevent OOM
+        // caused by the excessive data of the last snapshot split.
+        remainingSplits.add(0, schemaLessSnapshotSplits.get(size - 1));
+        remainingSplits.addAll(schemaLessSnapshotSplits.subList(0, size - 1));
```
Comment (gong) [inline inlong-sort/sort-flink/sort-flink-v1.13/sort-connectors/mysql-cdc/src/main/java/org/apache/inlong/sort/cdc/mysql/source/assigners/MySqlSnapshotSplitAssigner.java:405]: `remainSplits` is all splits, `remainingSplits.add(0, schemaLessSnapshotSplits.get(size - 1));` will cause other table split lost

Comment (e-mhui) [reply inlong-sort/sort-flink/sort-flink-v1.13/sort-connectors/mysql-cdc/src/main/java/org/apache/inlong/sort/cdc/mysql/source/assigners/MySqlSnapshotSplitAssigner.java:405]: The `remainingSplits.add(0, schemaLessSnapshotSplits.get(size - 1));` means  Inserts the specified element at the specified position in this list, it's not cause other split lost.

**Review by gong — APPROVED** [review]

**Pull request merged**